### DISCUSSION TITLE
Use classes for components

### DISCUSSION THREAD

**Discussion by masc-it** [pyedifice/pyedifice#186] [Ideas]
Hey, 

first thing first, great job with edifice, I think it's a great attempt to have a usable and scalable GUI library in py-land. State management in vanilla qt gets horrible the bigger the app gets, as you well know.

The first thing that got me when I first used pyedifice is the ubiquitous use of functions. Of course I know that this is inspired from react and in js-land is very common to "just use functions", but in py-land, well, there are more suitable constructs imho.

Take this snippet (from an example):

```
@component
def MyComponent(self, start: int):
    child_makers = []
    for i in range(start, start + 100):

        def child_maker(i_):
            def _child_maker():
                return Label(text=str(i_), style={"width": 50}).set_key(str(i_))

            return _child_maker

        child_makers.append(
            FixedChildMaker(
                50,
                child_maker(i),
            )
        )
    FullRowViewFixed(child_makers=child_makers)
```

The main question I have is: Why the components are modeled as functions? 

For me it's kinda natural to see it modeled as some class that extends some base component (or + a dataclass with parameters). 

Also, for the user is kinda confusing seeing a "self" in a global function. In the docstrings it is also stated that is needed for internals, so it is objectively of no use for the end user. This reinforces the fact that a function with a component decorator is not the best engineering choice. Or maybe it can be done in a different way, without exposing the self?

Also, you should take into account that having too many nested functions is not good for your stack, since you're going to allocate a lot of stuff at every invocation.

Of course, take this as a constructive feedback, I see a lot of potential in this project!

Feel free to expand and give context, I am curious to understand the history of pyedifice as well, I know you got your reasons to engineer it in this way :) 💯

**Comment by jamesdbrock** [comment]
Hi @masc-it , thanks for writing.

> State management in vanilla qt gets horrible the bigger the app gets, as you well know.

Yeah that was exactly my motivation for using Edifice.

> The first thing that got me when I first used pyedifice is the ubiquitous use of functions. Of course I know that this is inspired from react and in js-land is very common to "just use functions", but in py-land, well, there are more suitable constructs imho.
> 
> Take this snippet (from an example):

Oh yeah don’t refer to that example, I should delete it. That is an old example from before Edifice had the [special `children` prop](https://pyedifice.github.io/stubs/edifice.component.html#composition) and all of that function indirection is a workaround for not having the special `children` prop.

> The main question I have is: Why the components are modeled as functions?
> 
> For me it's kinda natural to see it modeled as some class that extends some base component (or + a dataclass with parameters).

Pure functions and plain data feel like the most natural way of programming to me. Object inheritance was always a [“fundamentally flawed mechanism”](http://www.stlport.org/resources/StepanovUSA.html) and most object-oriented programming language communities these days advise avoiding that part of the language. In retrospect that makes sense. When do we actually need to do dynamic function dispatch per-instance? Almost never. Function dispatch on classes of types is almost always what we need.

> Also, for the user is kinda confusing seeing a "self" in a global function. In the docstrings it is also stated that is needed for internals, so it is objectively of no use for the end user. This reinforces the fact that a function with a component decorator is not the best engineering choice. Or maybe it can be done in a different way, without exposing the self?

Yeah that's bad and I and others have tried to remove that `self` but it's tricky. I should try again.

> Also, you should take into account that having too many nested functions is not good for your stack, since you're going to allocate a lot of stuff at every invocation.

That was good advice 30 years ago but I don't think it matters much any more. I've never blown the Python stack in non-recursive functions. Though “stackless Python” exists so I guess someone has had that problem?

> I am curious to understand the history of pyedifice as well, I know you got your reasons to engineer it in this way :) 💯

David Ding started Edifice in 2021 and he got the basic functionality working with an API that was like old React with class components. Then he abandoned the project and I and some colleagues adopted it and [upgraded to function components and Hooks](https://pyedifice.github.io/versions.html#v0-2-0) like React did in version 16.8.

Fundamentally I enjoy the React API style of function components and Hooks. In particular I like how Hooks solve the problems of [state, lifecycle methods, and sharing non-visual logic](https://ui.dev/why-react-hooks#react-hooks). And I deliberately try to imitate the React API so that Edifice users can benefit from their React experience.
